File: hto/33/hto_33.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <stdbool.h>
#include <math.h>
#include <stdint.h>

bool recur(int *array, int i, int j) {
    if (i >= j) return true;
    int p = i;
    while (array[p] < array[j]) p++;
    int m = p;
    while (array[p] > array[j]) p++;
    return p == j && recur(array, i, m - 1) && recur(array, m, j - 1);
}

bool verifyPostorder(int* postorder, int postorderSize){
    return recur(postorder, 0, postorderSize - 1);
}


int main(void) {
    int a[] = {1, 3, 2, 6, 5};
    printf("%s", verifyPostorder(a, 5) ? "true" : "false");
}

File: hto/33/hto_33.py
from typing import List, Dict
from queue import Queue
from collections import deque
import math
import itertools
    
class Solution:
    def verifyPostorder(self, postorder: [int]) -> bool:
        stack, root = [], float("+inf")
        for i in reversed(range(len(postorder))):
            if (postorder[i] > root):
                return False
            while (stack and postorder[i] < stack[-1]):
                root = stack.pop()
            stack.append(postorder[i])
        return True


if __name__ == "__main__":
    post = [4, 8, 6, 12, 16, 14, 10]
    s = Solution()
    print(s.verifyPostorder(post))
